File: src/lib/research/planner.ts
import type {
  CoverageFamily,
  CoverageFamilyStatus,
  RegulatoryAngle,
  ResearchHypothesis,
  ResearchTask,
  ScopePack
} from "./types";
import { blockedToolIdsForRole, researchWorkerToolIds } from "./toolCatalog";
import { PROGRAM_REGISTRY, type ProgramRegistryEntry } from "./programRegistry";
import { jurisdictionContextFor } from "./jurisdictionResolve";

// The planner generates a hypothesis TASK LIST from the program registry — the
// single source of truth — instead of a hardcoded family array or a fixed pool
// of angles. For each registry program whose triggeredBy(scope) fires (or whose
// family an SDS review flagged), every hypothesis on that program becomes a
// research task. Adding a program to the registry adds it to the plan; nothing
// here is hardcoded per family or per hypothesis id.
export function planResearch(scope: ScopePack, sdsActiveFamilies: ReadonlySet<CoverageFamily> = new Set()) {
  // Families come from the registry, not a hardcoded list.
  const families = [...new Set(PROGRAM_REGISTRY.map((p) => p.family))];
  const coverage_family_statuses = families.map((family) =>
    coverageStatusFor(family, scope, sdsActiveFamilies.has(family))
  );

  const activeFamilies = new Set(
    coverage_family_statuses.filter((s) => s.status !== "out_of_scope").map((s) => s.family)
  );

  // A program is in-scope when its own trigger fires OR an SDS flagged its family
  // (and the family wasn't ruled out_of_scope). One regulatory angle per program.
  const activePrograms = PROGRAM_REGISTRY.filter(
    (program) =>
      activeFamilies.has(program.family) && (program.triggeredBy(scope) || sdsActiveFamilies.has(program.family))
  );

  const familyStatusBy = new Map(coverage_family_statuses.map((s) => [s.family, s]));
  const regulatory_angles: RegulatoryAngle[] = activePrograms.map((program) =>
    angleForProgram(program, familyStatusBy.get(program.family))
  );

  // Hypotheses are the task list: every hypothesis of every active program.
  const research_graph: ResearchHypothesis[] = activePrograms.flatMap((program) =>
    program.hypotheses.map((h) => hypothesisFromRegistry(program, h, familyStatusBy.get(program.family)))
  );
  // Resolve the local jurisdiction once and hand every research task the same
  // orienting context (controlling authorities + local skill bodies + gaps).
  const jurisdiction_context = jurisdictionContextFor(scope.facility);
  const research_tasks = research_graph.map((h) => taskForHypothesis(h, jurisdiction_context));

  return { coverage_family_statuses, regulatory_angles, research_graph, research_tasks };
}

// Per-family in-scope / blocked / out-of-scope status. This is fact-driven (not a
// hardcoded pool): each family reads the scope facts its programs depend on and
// reports missing facts that block a confident determination.
function coverageStatusFor(family: CoverageFamily, scope: ScopePack, sdsFlagged: boolean): CoverageFamilyStatus {
  const equipmentKinds = scope.project_change.equipment.map((item) => item.kind);
  const hasChemicals = scope.project_change.chemicals.length > 0;
  const hasWaste = scope.project_change.waste_streams.length > 0;
  const disturbance = scope.project_change.disturbance_acres;
  const id = `CF-${family.toUpperCase()}`;

  if (family === "air") {
    const equipmentActive = scope.project_change.equipment.length > 0;
    const active = equipmentActive || sdsFlagged;
    return {
      id,
      family,
      status: active ? "active" : "out_of_scope",
      reason: equipmentActive
        ? "Project adds equipment that may emit air contaminants."
        : sdsFlagged
          ? "SDS review flagged VOC or air-emissions relevance; air permit applicability requires review."
          : "No equipment added that could emit air contaminants.",
      project_facts_considered: sdsFlagged ? [...equipmentKinds, "sds:voc_air_emissions_review"] : equipmentKinds,
      missing_facts: []
    };
  }

  if (family === "stormwater") {
    const missingCode = !scope.facility.sic && !scope.facility.naics && disturbance === null;
    return {
      id,
      family,
      status: missingCode ? "blocked_missing_fact" : "active",
      reason: missingCode
        ? "SIC/NAICS and disturbance acreage are missing."
        : "Industrial activity codes or construction acreage require stormwater review.",
      project_facts_considered: [`sic=${scope.facility.sic}`, `naics=${scope.facility.naics}`, `acres=${disturbance}`],
      missing_facts: missingCode ? ["facility.naics_or_sic", "project_change.disturbance_acres"] : []
    };
  }

  if (family === "hazmat") {
    const missingQuantity = hasChemicals && scope.project_change.chemicals.some((chemical) => chemical.quantity === null);
    return {
      id,
      family,
      status: !hasChemicals ? (sdsFlagged ? "active" : "out_of_scope") : missingQuantity ? "blocked_missing_fact" : "active",
      reason: hasChemicals
        ? "Project includes hazardous material storage."
        : sdsFlagged
          ? "SDS review flagged hazardous material content; HMBP applicability requires review."
          : "No hazardous materials indicated in intake.",
      project_facts_considered: scope.project_change.chemicals.map((chemical) => `${chemical.name}:${chemical.quantity ?? "missing"} ${chemical.unit ?? ""}`),
      missing_facts: missingQuantity ? ["chemicals.quantity", "chemicals.unit"] : []
    };
  }

  if (family === "waste") {
    return {
      id,
      family,
      status: hasWaste || sdsFlagged ? "active" : "out_of_scope",
      reason: hasWaste
        ? "Project identifies waste streams that need generator-status review."
        : sdsFlagged
          ? "SDS review flagged hazardous waste relevance; generator-status review required."
          : "No waste stream indicated.",
      project_facts_considered: scope.project_change.waste_streams.map((stream) => `${stream.description}:${stream.kg_per_month ?? "missing"} kg/month`),
      missing_facts: scope.project_change.waste_streams.some((stream) => stream.kg_per_month === null)
        ? ["waste_streams.kg_per_month"]
        : []
    };
  }

  if (family === "wastewater") {
    const discharge = scope.project_change.process_discharge;
    return {
      id,
      family,
      status: discharge === null ? (sdsFlagged ? "active" : "blocked_missing_fact") : discharge ? "active" : sdsFlagged ? "active" : "out_of_scope",
      reason:
        discharge === null
          ? sdsFlagged
            ? "SDS review flagged spill/stormwater containment relevance; pretreatment review required."
            : "Process discharge status is missing."
          : discharge
            ? "Project may discharge process wastewater."
            : "No process wastewater discharge indicated.",
      project_facts_considered: [`process_discharge=${discharge}`],
      missing_facts: discharge === null ? ["project_change.process_discharge"] : []
    };
  }

  // Any future family in the registry with no special fact logic yet: active when
  // its programs trigger or an SDS flags it; otherwise out_of_scope. Fail-open to
  // active (a new family should be investigated, not silently skipped).
  return {
    id,
    family,
    status: sdsFlagged ? "active" : "active",
    reason: "Registry program applies to this family; investigate.",
    project_facts_considered: [],
    missing_facts: []
  };
}

function angleForProgram(program: ProgramRegistryEntry, status: CoverageFamilyStatus | undefined): RegulatoryAngle {
  return {
    id: `A-${program.id}`,
    family: program.family,
    label: program.name,
    reason: program.what_it_does,
    triggering_facts: status?.project_facts_considered ?? [],
    status: status?.status ?? "active"
  };
}

function hypothesisFromRegistry(
  program: ProgramRegistryEntry,
  h: ProgramRegistryEntry["hypotheses"][number],
  status: CoverageFamilyStatus | undefined
): ResearchHypothesis {
  return {
    id: h.id,
    angle_id: `A-${program.id}`,
    family: program.family,
    question: h.question,
    claim_to_test: h.claim_to_test,
    required_facts: status?.project_facts_considered ?? [],
    expected_source_type: "regulation",
    success_criteria: [
      "official or high-authority source",
      "quote contains trigger, threshold, exemption, or blocker",
      "predicate evaluation is reproducible"
    ],
    dependencies: []
  };
}

export function taskForHypothesis(hypothesis: ResearchHypothesis, jurisdiction_context?: string): ResearchTask {
  return {
    task_id: `T-${hypothesis.id.slice(2)}`,
    hypothesis_id: hypothesis.id,
    assigned_agent: `${hypothesis.family}_researcher`,
    allowed_tools: researchWorkerToolIds(),
    blocked_tools: blockedToolIdsForRole("researcher"),
    budget: {
      // Headroom for genuine investigation: discover via web_search, fetch + corroborate
      // across multiple authorities, and ground — not just fetch-one-known-url.
      max_sources: 5,
      max_runtime_seconds: 90,
      max_model_calls: 8
    },
    ...(jurisdiction_context ? { jurisdiction_context } : {})
  };
}

File: src/research_core/planner.py
from __future__ import annotations

from collections.abc import Iterable
from typing import Any, Literal

from pydantic import BaseModel, model_serializer

from research_core.jurisdiction_resolve import jurisdiction_context_for
from research_core.models import ScopePack
from research_core.registry import PROGRAM_REGISTRY, ProgramHypothesis, ProgramRegistryEntry

CoverageStatus = Literal[
    "active",
    "blocked_missing_fact",
    "out_of_scope",
    "discovery_candidate",
]
ExpectedSourceType = Literal[
    "statute",
    "regulation",
    "agency_guidance",
    "permit_portal",
    "technical_doc",
]


class CoverageFamilyStatus(BaseModel):
    id: str
    family: str
    status: CoverageStatus
    reason: str
    project_facts_considered: list[str]
    missing_facts: list[str]


class RegulatoryAngle(BaseModel):
    id: str
    family: str
    label: str
    reason: str
    triggering_facts: list[str]
    status: CoverageStatus


class ResearchHypothesis(BaseModel):
    id: str
    angle_id: str
    family: str
    question: str
    claim_to_test: str | None = None
    required_facts: list[str]
    expected_source_type: ExpectedSourceType
    success_criteria: list[str]
    dependencies: list[str]


class ResearchTaskBudget(BaseModel):
    max_sources: int
    max_runtime_seconds: int
    max_model_calls: int


class ResearchTask(BaseModel):
    task_id: str
    hypothesis_id: str
    assigned_agent: str
    allowed_tools: list[str]
    blocked_tools: list[str]
    budget: ResearchTaskBudget
    jurisdiction_context: str | None = None

    @model_serializer(mode="wrap")
    def _serialize_without_absent_context(self, handler) -> dict[str, Any]:
        data = handler(self)
        if self.jurisdiction_context is None:
            data.pop("jurisdiction_context", None)
        return data


class Plan(BaseModel):
    coverage_family_statuses: list[CoverageFamilyStatus]
    regulatory_angles: list[RegulatoryAngle]
    research_graph: list[ResearchHypothesis]
    research_tasks: list[ResearchTask]


UNIVERSAL_TOOL_IDS = (
    "send_message",
    "emit_trace_event",
    "validate_artifact_schema",
    "log_step",
    "escalate_to_human",
)
RESEARCHER_CORE_TOOL_IDS = (
    "read_skill",
    "get_triggers",
    "get_source_pointers",
    "get_cached_source",
    "fetch_source",
    "prove_currency",
    "extract_threshold",
    "evaluate_predicate",
    "quarantine_injection",
    "analyze_voc_content",
    "verify_chemical_composition",
    "lookup_cas_hazards",
    "compute_aggregate_quantity",
)
BLOCKED_RESEARCHER_TOOL_IDS = (
    "get_form",
    "build_applicability_matrix",
    "generate_compliance_calendar",
    "assemble_review_package",
    "freshness_sweep",
    "propose_map_entry",
    "propose_form_entry",
)


def _unique(ids: Iterable[str]) -> list[str]:
    return list(dict.fromkeys(ids))


def research_worker_tool_ids() -> list[str]:
    return _unique((*UNIVERSAL_TOOL_IDS, *RESEARCHER_CORE_TOOL_IDS))


def blocked_tool_ids_for_role(role: str) -> list[str]:
    if role == "researcher":
        return list(BLOCKED_RESEARCHER_TOOL_IDS)
    return []


def plan_research(
    scope: ScopePack,
    sds_active_families: frozenset[str] | set[str] = frozenset(),
) -> Plan:
    families = _unique(program.family for program in PROGRAM_REGISTRY)
    coverage_family_statuses = [
        coverage_status_for(
            family,
            scope,
            family in sds_active_families,
        )
        for family in families
    ]
    active_families = {
        status.family
        for status in coverage_family_statuses
        if status.status != "out_of_scope"
    }
    active_programs = [
        program
        for program in PROGRAM_REGISTRY
        if program.family in active_families
        and (program.triggered_by(scope) or program.family in sds_active_families)
    ]
    family_status_by = {status.family: status for status in coverage_family_statuses}
    regulatory_angles = [
        angle_for_program(program, family_status_by.get(program.family))
        for program in active_programs
    ]
    research_graph = [
        hypothesis_from_registry(program, hypothesis, family_status_by.get(program.family))
        for program in active_programs
        for hypothesis in program.hypotheses
    ]
    jurisdiction_context = (
        jurisdiction_context_for(
            {"county": scope.facility.county, "city": scope.facility.city}
        )
        if scope.facility.county or scope.facility.city
        else None
    )
    research_tasks = [
        task_for_hypothesis(hypothesis, jurisdiction_context)
        for hypothesis in research_graph
    ]
    return Plan(
        coverage_family_statuses=coverage_family_statuses,
        regulatory_angles=regulatory_angles,
        research_graph=research_graph,
        research_tasks=research_tasks,
    )


def coverage_status_for(
    family: str,
    scope: ScopePack,
    sds_flagged: bool,
) -> CoverageFamilyStatus:
    equipment_kinds = [item.kind for item in scope.project_change.equipment]
    has_chemicals = len(scope.project_change.chemicals) > 0
    has_waste = len(scope.project_change.waste_streams) > 0
    disturbance = scope.project_change.disturbance_acres
    id = f"CF-{family.upper()}"

    if family == "air":
        equipment_active = len(scope.project_change.equipment) > 0
        active = equipment_active or sds_flagged
        return CoverageFamilyStatus(
            id=id,
            family=family,
            status="active" if active else "out_of_scope",
            reason=(
                "Project adds equipment that may emit air contaminants."
                if equipment_active
                else (
                    "SDS review flagged VOC or air-emissions relevance; air "
                    "permit applicability requires review."
                    if sds_flagged
                    else "No equipment added that could emit air contaminants."
                )
            ),
            project_facts_considered=(
                [*equipment_kinds, "sds:voc_air_emissions_review"]
                if sds_flagged
                else equipment_kinds
            ),
            missing_facts=[],
        )

    if family == "stormwater":
        missing_code = (
            not scope.facility.sic
            and not scope.facility.naics
            and disturbance is None
        )
        return CoverageFamilyStatus(
            id=id,
            family=family,
            status="blocked_missing_fact" if missing_code else "active",
            reason=(
                "SIC/NAICS and disturbance acreage are missing."
                if missing_code
                else (
                    "Industrial activity codes or construction acreage require "
                    "stormwater review."
                )
            ),
            project_facts_considered=[
                f"sic={scope.facility.sic}",
                f"naics={scope.facility.naics}",
                f"acres={disturbance}",
            ],
            missing_facts=(
                ["facility.naics_or_sic", "project_change.disturbance_acres"]
                if missing_code
                else []
            ),
        )

    if family == "hazmat":
        missing_quantity = has_chemicals and any(
            chemical.quantity is None
            for chemical in scope.project_change.chemicals
        )
        return CoverageFamilyStatus(
            id=id,
            family=family,
            status=(
                "active"
                if (not has_chemicals and sds_flagged)
                else (
                    "out_of_scope"
                    if not has_chemicals
                    else "blocked_missing_fact"
                    if missing_quantity
                    else "active"
                )
            ),
            reason=(
                "Project includes hazardous material storage."
                if has_chemicals
                else (
                    "SDS review flagged hazardous material content; HMBP "
                    "applicability requires review."
                    if sds_flagged
                    else "No hazardous materials indicated in intake."
                )
            ),
            project_facts_considered=[
                f"{chemical.name}:{chemical.quantity if chemical.quantity is not None else 'missing'} {chemical.unit or ''}"
                for chemical in scope.project_change.chemicals
            ],
            missing_facts=(
                ["chemicals.quantity", "chemicals.unit"] if missing_quantity else []
            ),
        )

    if family == "waste":
        return CoverageFamilyStatus(
            id=id,
            family=family,
            status="active" if has_waste or sds_flagged else "out_of_scope",
            reason=(
                "Project identifies waste streams that need generator-status review."
                if has_waste
                else (
                    "SDS review flagged hazardous waste relevance; generator-status "
                    "review required."
                    if sds_flagged
                    else "No waste stream indicated."
                )
            ),
            project_facts_considered=[
                f"{stream.description}:{stream.kg_per_month if stream.kg_per_month is not None else 'missing'} kg/month"
                for stream in scope.project_change.waste_streams
            ],
            missing_facts=(
                ["waste_streams.kg_per_month"]
                if any(
                    stream.kg_per_month is None
                    for stream in scope.project_change.waste_streams
                )
                else []
            ),
        )

    if family == "wastewater":
        discharge = scope.project_change.process_discharge
        return CoverageFamilyStatus(
            id=id,
            family=family,
            status=(
                "active"
                if discharge or sds_flagged
                else "blocked_missing_fact"
                if discharge is None
                else "out_of_scope"
            ),
            reason=(
                (
                    "SDS review flagged spill/stormwater containment relevance; "
                    "pretreatment review required."
                )
                if discharge is None and sds_flagged
                else "Process discharge status is missing."
                if discharge is None
                else "Project may discharge process wastewater."
                if discharge
                else "No process wastewater discharge indicated."
            ),
            project_facts_considered=[f"process_discharge={discharge}"],
            missing_facts=(
                ["project_change.process_discharge"] if discharge is None else []
            ),
        )

    return CoverageFamilyStatus(
        id=id,
        family=family,
        status="active",
        reason="Registry program applies to this family; investigate.",
        project_facts_considered=[],
        missing_facts=[],
    )


def angle_for_program(
    program: ProgramRegistryEntry,
    status: CoverageFamilyStatus | None,
) -> RegulatoryAngle:
    return RegulatoryAngle(
        id=f"A-{program.id}",
        family=program.family,
        label=program.name,
        reason=program.what_it_does,
        triggering_facts=status.project_facts_considered if status else [],
        status=status.status if status else "active",
    )


def hypothesis_from_registry(
    program: ProgramRegistryEntry,
    hypothesis: ProgramHypothesis,
    status: CoverageFamilyStatus | None,
) -> ResearchHypothesis:
    return ResearchHypothesis(
        id=hypothesis.id,
        angle_id=f"A-{program.id}",
        family=program.family,
        question=hypothesis.question,
        claim_to_test=hypothesis.claim_to_test,
        required_facts=status.project_facts_considered if status else [],
        expected_source_type="regulation",
        success_criteria=[
            "official or high-authority source",
            "quote contains trigger, threshold, exemption, or blocker",
            "predicate evaluation is reproducible",
        ],
        dependencies=[],
    )


def task_for_hypothesis(
    hypothesis: ResearchHypothesis,
    jurisdiction_context: str | None = None,
) -> ResearchTask:
    return ResearchTask(
        task_id=f"T-{hypothesis.id[2:]}",
        hypothesis_id=hypothesis.id,
        assigned_agent=f"{hypothesis.family}_researcher",
        allowed_tools=research_worker_tool_ids(),
        blocked_tools=blocked_tool_ids_for_role("researcher"),
        budget=ResearchTaskBudget(
            # Headroom for the real flow: read_skill -> web_search -> web_fetch/browser
            # -> read + corroborate -> submit_finding, with room to recover from a bad
            # fetch. 4 turns was too tight (couldn't even reach submit after orienting).
            max_sources=5,
            max_runtime_seconds=120,
            max_model_calls=10,
        ),
        jurisdiction_context=jurisdiction_context,
    )
